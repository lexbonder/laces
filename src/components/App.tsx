import { useState } from 'react';
import Container from 'react-bootstrap/Container';
import Nav from 'react-bootstrap/Nav';
import Navbar from 'react-bootstrap/Navbar';

import CharacterName from './CharacterName';
import DoSkillModal from './Modals/DoSkillModal';
import NewSkillModal from './Modals/NewSkillModal';
import SkillList from './SkillList';
import ManageXPButton from './ManageXPButton';
import AddSkillButton from './AddSkillButton';
import RootProvider from './RootProvider';
import XPModal from './Modals/XPModal';
import { Skill } from '../store/skill-context';

const App = () => {
    const [showDoSkillModal, setShowDoSkillModal] = useState(false);
    const [activeSkill, setActiveSkill] = useState<Skill | null>(null);
    const [showNewSkillModal, setShowNewSkillModal] = useState(false);
    const [showXPModal, setShowXPModal] = useState(false);

    const openDoSkillModal = (chosenSkill: Skill) => {
        setActiveSkill(chosenSkill);
        setShowDoSkillModal(true);
    };

    const resetActiveSkill = () => {
        setActiveSkill(null);
    };

    return (
        <RootProvider>
            <div className="bg-light" style={{ minHeight: '100vh' }}>
                <Navbar sticky="top" expand="lg" className="bg-light">
                    <Container fluid>
                        <CharacterName />
                        <Navbar.Brand className="text-center me-0" style={{ flex: 1 }}>
                            <h1 className="mb-0">Laces</h1>
                            <p className="h6 mb-0">
                                The unofficial{' '}
                                <a href="https://rollforshoes.com" title="Roll for Shoes official site">
                                    Roll for Shoes
                                </a>{' '}
                                companion app
                            </p>
                        </Navbar.Brand>
                        <Nav className="justify-content-end" style={{ flex: 1 }}>
                            <AddSkillButton showNewSkillModal={() => setShowNewSkillModal(true)} />
                            <ManageXPButton toggleXPModal={() => setShowXPModal(!showXPModal)} />
                        </Nav>
                    </Container>
                </Navbar>
                <Container fluid className="px-0">
                    <main className="flex-fill">
                        <SkillList openSkillModal={openDoSkillModal} />
                    </main>
                </Container>
                <XPModal show={showXPModal} hide={() => setShowXPModal(false)} />
                <NewSkillModal
                    activeSkill={activeSkill}
                    show={showNewSkillModal}
                    hide={() => setShowNewSkillModal(false)}
                    resetActiveSkill={resetActiveSkill}
                />
                {activeSkill && (
                    <DoSkillModal
                        activeSkill={activeSkill}
                        show={showDoSkillModal}
                        hide={() => setShowDoSkillModal(false)}
                        openNewSkillModal={() => setShowNewSkillModal(true)}
                        resetActiveSkill={resetActiveSkill}
                    />
                )}
            </div>
        </RootProvider>
    );
};

export default App;
