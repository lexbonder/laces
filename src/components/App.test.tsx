import { render, screen } from '@testing-library/react';
import App from './App';
import userEvent from '@testing-library/user-event';

describe('App', () => {
    test('renders the App title', () => {
        render(<App />);

        const title = screen.getByRole('heading', { name: 'Laces' });
        expect(title).toBeInTheDocument();
    });

    test('render the DoSkillModal', async () => {
        render(<App />);

        const doAnythingButton = screen.getByRole('button', { name: 'Do Anything' });
        await userEvent.click(doAnythingButton);

        const doSkillModal = screen.getByRole('dialog', { name: 'Do Anything 1' });
        expect(doSkillModal).toBeInTheDocument();

        const closeButton = screen.getByRole('button', { name: 'Close' });
        await userEvent.click(closeButton);

        expect(doSkillModal).not.toBeInTheDocument();
    });

    test('render the NewSkillModal', async () => {
        render(<App />);

        const newSkillButton = screen.getByRole('button', { name: 'Add Skill' });
        await userEvent.click(newSkillButton);

        const newSkillModal = screen.getByRole('dialog', { name: 'Add New Skill' });
        expect(newSkillModal).toBeInTheDocument();

        const closeButton = screen.getByRole('button', { name: 'Close' });
        await userEvent.click(closeButton);

        expect(newSkillModal).not.toBeInTheDocument();
    });

    test('render the ManageXPModal', async () => {
        render(<App />);

        const manageXPButton = screen.getByRole('button', { name: /xp:/i });
        await userEvent.click(manageXPButton);

        const xpModal = screen.getByRole('dialog', { name: 'Manage XP' });
        expect(xpModal).toBeInTheDocument();

        const closeButton = screen.getByRole('button', { name: 'Close' });
        await userEvent.click(closeButton);

        expect(xpModal).not.toBeInTheDocument();
    });
});
