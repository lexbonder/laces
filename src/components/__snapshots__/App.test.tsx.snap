// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders the App component 1`] = `Screen {}`;
